Add Identity Center Account Assignments to Unified Resource Cache

tcsc opened the pull request:
Adds Identity Center Account Assignments to the Unified resource cache so they can be requested in access requests.

Unfortunately we can't just include an `identitycenterv1.AccountAssignment` directly in the resource cache `ListResources` output because the legacy protobuf codegen used for the `authservice` and the new codegen used for `identitycenter/v1` produce incompatible serialization code, so resulting generated code does not compile.

To get around this issue, this change introduces a parallel (and slightly simplified) definition of an `IdentityCenterAccountAssignment` in the `authservice` protobuf spec to act as the wire format for this type. The cached `identitycenterv1.AccountAssignment` resources are copied into a  `proto.IdentityCenterAccountAssignment` on a cache read. 

I'm not particularly thrilled with this solution, and if anyone has an alternative that lets us use the definitions from `identitycenter/v1` without this parallel definition, I'd be much obliged.

Change Includes:
 - adding resources to cache
 - adding account assignment paginated resource
 - AccountAssignment Role Condition matching for RBAC

r0mant reviewed (commented):
@tcsc Does this need a rebase? Feels like it contains code from PRs I already reviewed.

tcsc commented on api/proto/teleport/legacy/client/proto/authservice.proto:
```diff
@@ -1934,6 +1934,42 @@ message CreateRegisterChallengeRequest {
+
+// IdentityCenterAccountAssignment represents a requestable Identity Center
+// Account Assignment
+message IdentityCenterAccountAssignment {
```
So, `types.IdentityCenterAccountAssignment` is for specifying Account Assignments in role conditions. If I had known bettrer back when defining this, I'd have called them an `AccountAssignmentCondition` or `AccountAssignmentExpression` (as they can be globbed).

This definition, `proto.IdentityCenterAccountAssignment` allows Account Assignment resources to be included in a list of resources returned by `ListResources` ( essentially by having a parallel (and slightly simplified) definition of an `identitycenterv1.AccountAssignment` in the legacy authservice protobuf definitions).

The primary `identitycenterv1.AccountAssignment` gets copied into a `proto.IdentityCenterAccountAssignment` at read time and then gets shipped over to whatever client invoked `ListResources().

The reason we can't just include an `identitycenterv1.AccountAssignment` in the `ListResources` output that the legacy protobuf codegen used for the `authservice` and the new codegen used for `identitycenter/v1` produce incompatible serialization code, so resulting generated code does not compile.

If there is a way around this that will let us use the definitions from `identitycenter/v1`, I'd be much happier.

tcsc commented:
> @tcsc Does this need a rebase? Feels like it contains code from PRs I already reviewed.

And how.

r0mant commented on api/client/client.go:
```diff
@@ -3822,6 +3826,26 @@ func (c *Client) ListResources(ctx context.Context, req proto.ListResourcesReque
 
+// repackIdentityCenterAccountAssignment converts a wire-format IdentityCenterAccountAssignment
+// resource back into an identitycenterv1.AccountAssignment instance.
+func repackIdentityCenterAccountAssignment(src *proto.IdentityCenterAccountAssignment) types.ResourceWithLabels {
```
Nit: Just a thought, can this maybe be moved to a more appropriate package where IC resources (or maybe `types`) and named `ICAccountAssignmentToResourceWithLabels`? We probably have other similar converters somewhere already, right? Like there's `types.LegacyTo153Metadata` you're using below.

Not a blocker though.

r0mant commented on api/proto/teleport/legacy/client/proto/authservice.proto:
```diff
@@ -1940,6 +1940,50 @@ message CreateRegisterChallengeRequest {
+// IdentityCenterAccount holds information about an Identity Center account
+// within an IdentityCenterAccountAssignment
+message IdentityCenterAccount {
+  string ID = 1;
```
Nit: Godocs please on all exported fields, here and below.

tcsc committed:
Add Identity Center Account Assignments to Unified Resource Cache

Adds Identity Center Account Assignments to the Unified resource cache
so they can be requested in access requests.

Unfortunately we can't just include an `identitycenterv1.AccountAssignment`
directly in the resource cache ListResources output because the legacy
protobuf codegen used for the authservice and the new codegen used for
identitycenter/v1 produce incompatible serialization code, so resulting
generated code does not compile.

To get around this issue, this change introduces a parallel (and slightly
simplified) definition of an IdentityCenterAccountAssignment in the authservice
protobuf spec to act as the wire format for this type. The cached
`identitycenterv1.AccountAssignment` resources are copied into a
`proto.IdentityCenterAccountAssignment` on a cache read.

Includes:
 - adding resources to cache
 - adding account assignment paginated resource
 - account assignment role condition matching for RBAC

tcsc committed:
Merge branch 'master' into tcsc/idc-expose-account-assignmens-via-resource-cache

tcsc committed:
Merge branch 'master' into tcsc/idc-expose-account-assignmens-via-resource-cache
